### ISSUE TITLE
mpi-python3 appears to be bust in string (unicode) vs bytes

### ISSUE THREAD

**Issue by xnox** [boostorg/mpi#111]
```
$ cat demo2.py 
# https://www.boost.org/doc/libs/1_67_0/doc/html/mpi/tutorial.html#mpi.python

import boost.mpi as mpi

if mpi.world.rank == 0:
    mpi.world.send(1, 0, 'Hello')
    msg = mpi.world.recv(1, 1)
    print(msg, '!')
else:
    msg = mpi.world.recv(0, 0)
    print((msg + ', '), end=' ')
    mpi.world.send(0, 1, 'world')
```

```
$ mpirun --oversubscribe --allow-run-as-root -np 2 python3 ./demo2.py
Traceback (most recent call last):
  File "./demo2.py", line 6, in <module>
    mpi.world.send(1, 0, 'Hello')
TypeError: Expecting an object of type str; got an object of type bytes instead
```

It seems like Python unicode string is encoded into bytes, instead of being processed as a Python unicode string?!

**Comment by xnox** [comment]
Hello, any update on this?

**Comment by aminiussi** [comment]
Not yet, I'll try to work on it this afternoon.

Thanks for the update

**Comment by xnox** [comment]
> It's not a matter of string. I can't get the test working with int either, and it's before reaching the c++ send code...

that can't be good

**Comment by aminiussi** [comment]
it could be a problem with my environment though.

**Comment by aminiussi** [comment]
it could be a problem with my environment though.

**Comment by aminiussi** [comment]
Ok, I'm got some "expected" problem. The point I mention above was due to the activation of my "production" boost distribution.

**Comment by aminiussi** [comment]
Still, it's the first time I look into this area (boost mpi python bindings) so there is a learning curve.

**Comment by aminiussi** [comment]
I'm starting to wonder if python3 support was even implemented.

**Comment by aminiussi** [comment]
Or maybe it's related with the fact that we now assume uniform cluster on default, and serialize the data in binary format in that case. This error occurs in pickle and such messages have been mentioned in that area.

**Comment by aminiussi** [comment]
I suspect it's related to: https://stackoverflow.com/questions/23582489/python-pickle-protocol-choice

Although it default to -1 in our case.

**Comment by aminiussi** [comment]
Or maybe:

```
str pickle::dumps(object obj, int protocol)
{
  if (!data) initialize_data();
  return extract<str>((data->dumps)(obj, protocol));
}
```
Which would only work with older python, since `str` does not look compatible with newer protocol which supports bytes.

**Comment by aminiussi** [comment]
```
[alainm@castor libs]$ python3
Python 3.7.4 (default, Nov 22 2019, 21:31:39) 
[GCC 7.3.0] :: Intel(R) Corporation on linux
Type "help", "copyright", "credits" or "license" for more information.
Intel(R) Distribution for Python is brought to you by Intel Corporation.
Please check out: https://software.intel.com/en-us/python-distibution
>>> import pickle
>>> pickle.dumps(42)
b'\x80\x03K*.'
>>> pickle.HIGHEST_PROTOCOL
4
>>> ^D
[alainm@castor libs]$ python2
Python 2.7.5 (default, Jul 13 2018, 13:06:57) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-28)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> pickle.dumps(42)
'I42\n.'
>>> pickle.HIGHEST_PROTOCOL
2
>>> 
[alainm@castor libs]$
```
Probably never worked in python3....

**Comment by aminiussi** [comment]
As for a start, in hotfix/pickle_py3_friendly_api, use object as buffer type.
Str is python2 specific, we need to deal with python3's bytes

**Comment by xnox** [comment]
> As for a start, in hotfix/pickle_py3_friendly_api, use object as buffer type.
> Str is python2 specific, we need to deal with python3's bytes

Pickled data is always binary, thus one can treat it as always bytes, in both python2 and python3.

Note that in python2 `byte == str` i.e. they are the same type. And just `byte` in python3.

The special class for text was `unicode` in python2 and simply `str` in python3.

**Comment by aminiussi** [comment]
> Pickled data is always binary, thus one can treat it as always bytes, in both python2 and python3.

python2 clearly encode it as a str object on default, although the protocol may chose a specific serialization technique.

Anyway, I'm starting the see the light (I hate debugging that section), should have something to test soon.

**Comment by xnox** [comment]
You say str object, but such thing does not exist in python2, it's bytes.

**Comment by aminiussi** [comment]
I have str on 2.7 and bites on 3.7
```
alainm@jarvis:~$ python2.7
Python 2.7.17 (default, Apr 15 2020, 17:20:14) 
[GCC 7.5.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> pickle.dumps(42)
'I42\n.'
>>> type(pickle.dumps(42))
<type 'str'>
>>> 
alainm@jarvis:~$ python3.6
Python 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> pickle.dumps(42)
b'\x80\x03K*.'
>>> type(pickle.dumps(42))
<class 'bytes'>
>>> 
alainm@jarvis:~$ 
```

**Comment by aminiussi** [comment]
Hi, 

If you can have a look at branch `hotfix/pickle_py3_friendly_api` ?

I still need to check that is still work with python 2 and then on my linux box.

It's been tested with Intel 2020 compiler,MPI,python3.7.4

**Comment by aminiussi** [comment]
Merged #121 

Still need to fix the module name issue

**Comment by aminiussi** [comment]
Buut that is another issue so we can consider trhis one closed
